### DISCUSSION TITLE
Williams Fractal

### DISCUSSION THREAD

**Discussion by DaveSkender** [DaveSkender/Stock.Indicators#255] [Indicators]
Community discussion about [Williams Fractal](https://dotnet.stockindicators.dev/indicators/Fractal/#content)

**Comment by 8AWESOME4** [comment]
Hi!
indicator Williams Fractal 
 bear and bull are always null
history 1 year

**Reply by DaveSkender** [reply]
What Symbol and timeframe are you using?  I'll take a look.
If you can provide your `history`, that would be helpful as well.

**Reply by DaveSkender** [reply]
I'm running it here on TSLA and not seeing any issues with the library, so there's probably an issue with `Close` price in your `history`.

![image](https://user-images.githubusercontent.com/8432125/104105472-52974600-527c-11eb-84c2-f1344435bdb5.png)

With that said, theoretically, you could get a resultset full of `nulls` if the Fractal pattern never occurs, but I think that would be rare.

**Comment by 8AWESOME4** [comment]
IDE VS 2019 NetFramework 4.8
Period 01/jan/2020 – 08/jan/2021
a pair of BTC / USDT interval 30 minutes
here is my code how i use
var RSI = Indicator.GetRsi(init_quote_200_Tick, 7).LastOrDefault().Rsi;
                var BB = Indicator.GetBollingerBands(init_quote_200_Tick).LastOrDefault();
                var Fractal = Indicator.GetFractal(init_quote_200_Tick).LastOrDefault();
                var bull = Fractal.FractalBull;
give results for all their properties, and the fractal returns null

History xml format 
[OutFullData.txt](https://github.com/DaveSkender/Stock.Indicators/files/5792111/OutFullData.txt)

**Reply by DaveSkender** [reply]
In your sample above, you're selecting `LastOrDefault`, which is only the last result value from the whole series.  It's always going to be null due to the way Fractal is calculated.  If you:

```csharp
var Fractal = Indicator.GetFractal(init_quote_200_Tick)
  .Where(x=>x.FractalBull != null || x.FractalBear != null)
  .LastOrDefault();
```
you're more likely to get the last known Fractal.  Keep in mind, the library returns an entire history of Fractal results, most of which will be null.  It will only have a value on dates when there is a Fractal.

**Reply by DaveSkender** [reply]
For my TSLA example, the last 15 periods in the result set would look like:

![image](https://user-images.githubusercontent.com/8432125/104114062-302e1880-52ce-11eb-84f4-06d3f41cb7d9.png)

Given the data you provided, your results should look something like: [results.txt](https://github.com/DaveSkender/Stock.Indicators/files/5792201/results.txt) (CSV)
(nulls omitted for clarity)

**Reply by 8AWESOME4** [reply]
Thank you !!! I apply example

**Comment by casperleenheer** [comment]
Hi,

I have the same sort of problem. I added my history. I keep track of assets and then per asset 5 lists of quotes with different intervals. If I run code like this:



```
var Fractal = Indicator.GetFractal(i.quoteList).Where(x => x.FractalBull != null || x.FractalBear != null).LastOrDefault();
if (Fractal != null)
{
    int stop = 0;
}
```


Of all the quotes I only get 1 hit. Are my quotes incorrect? I realized that for example the 5 minutes quotes are not spot on the 5 minutes, i.e. 06:07...06:12 instead of 06:05..06:10 etc. 

On tradingview the fractals do appear to pop up. I do get results for  
` var ema50 = i.quoteList.GetEma(50); `
so it looks like my quotes are ok.

Any help is really appreciated. 

[histquotes.zip](https://github.com/DaveSkender/Stock.Indicators/files/6982039/histquotes.zip)



Casper

**Reply by DaveSkender** [reply]
`.LastOrDefault()` will always return one and only one record.  If you want the whole series of results, do not use it

**Reply by casperleenheer** [reply]
I know, that's not the issue. The problem is that it will only find one hit out of the provided data set. The rest of the set is null. That's why I provided the dataset. I was hoping to get a hit for every set. I hope I made the issue more clear.

**Reply by DaveSkender** [reply]
Your dataset is not a single quote history, so it's going to be hard for me to reproduce what's happening.  I suspect this might have more to do with your wrapper code than this indicator library itself.  I can probably take a look at this, but you'll have to do a little work to make it easier on me.

Can you send the following:

1) The precise quote history that you think does not work (not a set of several quote histories)
2) The exact syntax, including paramter values, you are using on that precise quote history

**Reply by casperleenheer** [reply]
Hi,

Here is the code (using NewtonSoft)

```
string filenameread= "adaeur5m300.json";
List<Quote> myQuoteList = JsonConvert.DeserializeObject<List<Quote>>(File.ReadAllText(filenameread));
IEnumerable<FractalResult> resFractal = myQuoteList.GetFractal(2);
var FindFractal = resFractal.Where(x => x.FractalBull != null || x.FractalBear != null).LastOrDefault();
if (FindFractal != null)
{
    int stop = 0;
}
```

I attached the json. I hope this is what you want. Good luck and thanks!
[adaeur5m300.zip](https://github.com/DaveSkender/Stock.Indicators/files/6984340/adaeur5m300.zip)

**Reply by DaveSkender** [reply]
You are using `.LastOrDefault()`, which you say you know only produces one record.  I can tell you without doing any analysis, that it is only producing one record because you are using `.LastOrDefault()`.

Is this irrelevant code?  Or something you want me to consider?
```csharp
var FindFractal = resFractal.Where(x => x.FractalBull != null || x.FractalBear != null).LastOrDefault();
if (FindFractal != null)
{
    int stop = 0;
}
```

Can you also include the output of `resFractal`?

**Reply by casperleenheer** [reply]
Ok, here the adjusted code and the result file:

```
string filenameread = "adaeur5m300.json";
List<Quote> myQuoteList = JsonConvert.DeserializeObject<List<Quote>>(File.ReadAllText(filenameread));
IEnumerable<FractalResult> resFractal = myQuoteList.GetFractal(2);
using (StreamWriter sw = File.AppendText("result.txt"))
{
    foreach (var rrr in resFractal)
    {
        string line = "Bear: " + rrr.FractalBear + " Bull: " + rrr.FractalBull;
        Console.WriteLine(line);
        sw.WriteLine(line);
    }
}
```

[result.txt](https://github.com/DaveSkender/Stock.Indicators/files/6984835/result.txt)

**Reply by DaveSkender** [reply]
Thanks for the data; I'll take a look when I have more time.  For now, if you want to evaluate the `myQuoteList`, you can check to see if any Fractals exist.  With your span of 2, this means that a `FractalBear` would occur if the a High price is higher than the 2 High prices before and after.  Use lowest Low for `FractalBull`.  `Date` values are not considered, only sequence.

**Reply by casperleenheer** [reply]
I will do that and let you know.

**Reply by DaveSkender** [reply]
I misspoke and corrected by last statement, use High instead of Close for bull case.

**Reply by casperleenheer** [reply]
Sorry, but it is my data. I retrieve a ticker, but as I found out, the High and Low are of the day and not for the interval. Hence no result.

**Reply by DaveSkender** [reply]
I have an update to Williams Fractal that will allow the use of Close price, but it's not yet available in the library.  It's part of the modifications required for an upcoming Divergence Indicator - #434
